aQute.libg/src/aQute/libg/uri/URIUtil.java

```diff
@@ -41,4 +44,84 @@ public static URI resolve(URI baseURI, String reference) throws URISyntaxExcepti
 		return baseURI.resolve(refURI);
 	}
 
+	private static final URI EMPTYURI = URI.create("");
+
+	/**
+	 * Attempts to fetch a path on the file system for the given URI string.
+	 * Tries a few more tricks than the standard method of
+	 * <tt>Paths.get(new URI(uriString))</tt> - it can handle plain paths, and
+	 * also nested URI pseudo-schemes like <tt>reference:</tt> and
+	 * <tt>jar:</tt>. Examples:
+	 * <ul>
+	 * <li>/path/to/file => /path/to/file</li>
+	 * <li>reference:file:/path/to/file => /path/to/file</li>
+	 * <li>jar:file:/path/to/file.jar!/some/contained/element =>
+	 * /path/to/file.jar</li>
+	 * <li>http://server/path/to/file => <tt>null</tt></li>
+	 * </ul>
+	 *
+	 * @param uriString The string containing the URI for which we are
+	 *            attempting to construct a path.
+	 * @return The method's best guess as to which file on the local filesystem
+	 *         this URI refers, or <tt>null</tt> if it's invalid or not a local
+	 *         filesystem URI.
+	 */
+	public static Path fromURI(String uriString) {
+		if (uriString == null || uriString.isEmpty()) {
+			return null;
+		}
+
+		try {
+			URI uri = resolve(EMPTYURI, uriString);
+			while (uriString != null && !uriString.isEmpty()) {
+
+				final String scheme = uri.getScheme();
+				if (scheme == null) {
+					try {
+						return Paths.get(uriString);
+					} catch (InvalidPathException e) {
+						e.printStackTrace();
+						return null;
+					}
+				}
+
+				switch (uri.getScheme()
```

Comment:
See

https://github.com/bndtools/bnd/blob/672b366c691ebed6973cbb40a22f7abfe0c15e52/biz.aQute.bndlib/src/aQute/bnd/osgi/Resource.java#L36-L61

where we handle peeling off schemes (in a URL but basically similar). But the key thing in that code is to never use a `jar:file:` URL since the code in Java will lock the file in Windows for the duration of the VM.

I don't see that we need a loop since we are not going to see `jar:jar:jar:` for example. The worst case would seem to be `reference:jar:file`

Reply:
> But the key thing in that code is to never use a jar:file: URL since the code in Java will lock the file in Windows for the duration of the VM.

I understand this reasoning and I agree, but this method is not generating URIs - it is interpreting URIs that have been generated elsewhere. So I don't quite see the relevance of that comment to the current PR?

> I don't see that we need a loop since we are not going to see jar:jar:jar: for example. The worst case would seem to be reference:jar:file

I was trying to be as generic as possible without trying to pre-empt what all the possible combinations might be, so that it won't suddenly stop working in the future unexpectedly on some unforeseen combination. I can easily change it to handle only known valid combinations.